tinygrad/tensor.py

```diff
@@ -99,6 +99,9 @@ def __repr__(self):
   # Python has a non moving GC, so this should be okay
   def __hash__(self): return id(self)
 
+  def __reduce__(self) -> Tuple[Type, Tuple[List, Optional[Union[str, tuple, list]], Optional[DType], Optional[bool]]]:
+    return self.__class__, (self.data().tolist(), self.device, self.dtype, self.requires_grad)
```

Comment:
Does this handle multi-dimensional shapes? Can you add that to test?

Reply:
I have added `test_pickle_multi_dim` and `test_pickle_scalar` and both pass  :heavy_check_mark:

Reply:
Why is dtype optional? Why are you only pickling to CPU?

Reply:
`DType` was optional since one of the purposes of the `__reduce__` function is to return the arguments needed to reconstruct the `Tensor` object, meaning:

```py
y = Tensor.randn(shape)
x = Tensor(*y.__reduce__()[1])

np.testing.assert_equal(y.numpy(), x.numpy())
assert y.dtype == x.dtype
assert y.device == x.device
assert y.requires_grad == x.requires_grad

# (I have added this as a new test case in `test_tensor.py` - passes)
```

Since `dtype` is optional in `Tensor`'s constructor, `dtype` can be optional in `__reduce__`. However, I removed the type hints in my latest commits because the method signature alone would go over 2 lines with ruff's current max line length.

[Python Docs about reduce](https://docs.python.org/3/library/pickle.html#object.__reduce__)